### ISSUE TITLE
certbot does not handle orderNotReady finalize response properly

### ISSUE THREAD

**Issue by frasertweedale** [certbot/certbot#9766]
# Executive summary

If certbot attempts to finalizes the order and receives 403 (Forbidden) error with a problem document of type
"orderNotReady", certbot exits instead re-querying the order state to determine the next step.

Certbot should not exit abnormally and should instead re-query the order state and attempt finalization again.
Alternatively, Certbot should wait for the order status to transition to `ready` before attempting finalization.

# Context

Server implementations to not necessarily set authz status and order status atomically.  Certbot eagerly proceeds to finalization when all authz are valid, **regardless of the status of the order**.

This behaviour does not appear to violate the RFC, which is unfortunately vague and states ([Section 7.4](https://datatracker.ietf.org/doc/html/rfc8555#section-7.4):

>  Once the **client believes it has fulfilled the server's requirements**, it should send a POST request to the order resource's finalize URL.

If all authzs are `valid` but the order status is still (temporarily) `pending`, attempting to finalize will cause the server to respond 403 with the `orderNotReady` problem type.  The RFC specifies how a client should handle this scenario:

> A request to finalize an order will result in error if the order is not in the "ready" state.  In such cases, the server MUST return a 403 (Forbidden) error with a problem document of type "orderNotReady".  The client should then send a POST-as-GET request to the order resource to obtain its current state.  The status of the order will indicate what action the client should take (see below).
>
> ...
>
> -  "invalid": The certificate will not be issued.  Consider this order process abandoned.
> -  **"pending"**: The server does not believe that the client has fulfilled the requirements.  Check the "authorizations" array for entries that are still pending.
> - **"ready"**: The server agrees that the requirements have been fulfilled, and is awaiting finalization.  Submit a finalization request.
> - "processing": The certificate is being issued.  Send a POST-as-GET request after the time given in the Retry-After header field of the response, if any.
> - "valid": The server has issued the certificate and provisioned its URL to the "certificate" field of the order.  Download the certificate.

# Certbot's behaviour

Certbot terminates, instead of querying the order again as specified by the RFC.

This behaviour can be ovserved in this test run of the Red Hat Certificate System (Dogtag PKI) ACME server implementation: https://github.com/edewata/pki/actions/runs/6163037711/job/16726539917#step:28:169

# Expected behaviour

Certbot queries the order again.  One of two scenarios are likely:

- The order status is now `ready`, and certbot should attempt finalization again
- The order status is still `pending`, and certbot should review the authz statuses.  It will observe that they (as before) are all `valid`, and attempt to finalize again.

Alternatively, Certbot could be modified to **wait until the order status is `ready` before attempting finalization**, instead of eagerly proceeding to finalization when the authz are all `valid`.  This might be simpler to implement.  But it is conceivable that this approach could cause compatibility issues with some server implementations.  For example, if order status is left `pending` until the client asks to finalize the order.  However, such a server implementation explicitly violates [RFC 8555 section 7.1.6](https://www.rfc-editor.org/rfc/rfc8555.html#section-7.1.6):

```
   Order objects are created in the "pending" state.  Once all of the
   authorizations listed in the order object are in the "valid" state,
   the order transitions to the "ready" state.  ...
```

However, the **atomicity** of this transition with the transition of the final authz from `pending` to `valid` is not suggested or required by the RFC, and clients should not rely on such an assumption (as Certbot currently does).

**Comment by frasertweedale** [comment]
I haven't confirmed whether it's still happening but it most likely is, unless a developer deliberately fixed it.

**Comment by ohemorange** [comment]
It looks like this was noticed when used with the dogtag pki CA, and a fix was added on that end to avoid the issue in practice. Let's Encrypt, which certbot is used most often with, doesn't seem to trigger this issue. Given that, @frasertweedale, do you still have a need for this to be fixed?

**Comment by lennybacon** [comment]
Yes. This is still required. A pending certification request is part of the spec and used widely (especially in non-public environments). Therefore, it is part of the spec. 

The “workaround”, by omitting the pending state, mean lowering the security bar and is not acceptable for my scenarios and customers.

It would be really cool if certbot could act accordingly to the specified HTTP status codes.

**Comment by frasertweedale** [comment]
@ohemorange thanks for your attention to this issue.

Yes, we worked around this in Dogtag PKI by ensuring the order state transitions to "ready" before the final authorization state transitions to "valid" upon completion of the challenge (commit https://github.com/dogtagpki/pki/commit/2457fc36bfa4c7fa1cb7eb2d930ed1e98ac9b197).

But sooner or later, another ACME server implementation could come along with similar behaviour, and Certbot will fail in the same way.  The problem arises due to ambiguities in the spec.  However, the ACME spec is unambiguous about how clients should handle the `403 orderNotReady` response.  Certbot's current behaviour is non-conformant.  I don't think the issue should be closed or ignored just because particular server implementations avoid or worked around it.

As for @lennybacon's comment:

> The “workaround”, by omitting the pending state, mean lowering the security bar and is not acceptable for my scenarios and customers.

I don't understand how the workaround lowers the security.  The pending state still exists - the workaround manages the timing of the state transition differently, but the conditions for that transition are unchanged.  If I'm missing something could you please elaborate?

**Comment by lennybacon** [comment]
@frasertweedale unaware of your fix, I assumed you just skipped validation and the pending state which would have lowered the bar. You are right with your understanding.

**Comment by ohemorange** [comment]
I disagree that we're nonconformant, as it's a "should" and not a "MUST". But I do agree that it would be nice to have this functionality implemented, even if it's not currently causing problems.

**Comment by ohemorange** [comment]
Wait, are you sure this is still happening in the current certbot version? Here's the relevant portion of certbot's code:

https://github.com/certbot/certbot/blob/main/acme/acme/client.py#L195-L218

We do continue retrying/polling until we get status `valid`. We only error out if we specifically get the `invalid` status, which as per the RFC means "The certificate will not be issued.  Consider this order process abandoned."

> Once the client believes it has fulfilled the server's requirements, it should send a POST request to the order resource's finalize URL.

We do that in `acme.client.begin_finalization`.

> A request to finalize an order will result in error if the order is
>  not in the "ready" state.  In such cases, the server MUST return a
>    403 (Forbidden) error with a problem document of type
>    "orderNotReady".  The client should then send a POST-as-GET request
>    to the order resource to obtain its current state.  The status of the
>    order will indicate what action the client should take (see below).

We don't wait for the `ready` state, but then we don't need to, as mentioned in the previous citation. We immediately send the post-as-get as specified, and check the status. 

Are you reading the spec differently than we did, or looking at a previous version of certbot?

**Comment by ohemorange** [comment]
We do not error when the finalize request fails. We only error if we specifically get the status `invalid`, which is, per the spec, used to indicate the request will never be finalized, and should in fact fail. Where are you seeing that we error when the finalize request fails? If the server is responding with status `invalid` when it is in fact still `processing`, that is an error of the server implementation.

See this diagram from the RFC:
```


    pending --------------+
       |                  |
       | All authz        |
       | "valid"          |
       V                  |
     ready ---------------+
       |                  |
       | Receive          |
       | finalize         |
       | request          |
       V                  |
   processing ------------+
       |                  |
       | Certificate      | Error or
       | issued           | Authorization failure
       V                  V
     valid             invalid

                    State Transitions for Order Objects
```
We do not error on `pending`, `ready`, or `processing`, only `invalid`.

**Comment by ohemorange** [comment]
Perhaps you are saying that we should poll for `ready` status before sending the finalize request and then subsequently polling for final issuance? That is not required by the spec.

**Comment by ohemorange** [comment]
Ah, but perhaps you are talking about what happens before this, in `begin_finalization`:

https://github.com/certbot/certbot/blob/7322e56cc78afc526e2f18709856694ce2f39e0e/acme/acme/client.py#L220-L238

You are saying that if the state is not yet `ready`, that `post` requesting to finalize will return an error, and we do not check for that error. But we should not wait for the ready state: "Once the client believes it has fulfilled the server's requirements, it should send a POST request to the order resource's finalize URL." But the correct follow-up action is to "send a POST-as-GET request to the order resource to obtain its current state," which is in fact what we do in `poll_finalization`. We should not have to re-send that POST to the finalize URL.

Now, perhaps we error out if we receive that 403 in response to the post on line 236? In that case I think the correct solution is to wrap it in a `try` that catches an `orderNotReady` and ignores it.

**Comment by ohemorange** [comment]
So, based on my understanding, this would both address the failure you're seeing and be spec compliant:

```
    def finalize_order(self, orderr: messages.OrderResource, deadline: datetime.datetime,
                       fetch_alternative_chains: bool = False) -> messages.OrderResource:
        """Finalize an order and obtain a certificate.

        :param messages.OrderResource orderr: order to finalize
        :param datetime.datetime deadline: when to stop polling and timeout
        :param bool fetch_alternative_chains: whether to also fetch alternative
            certificate chains

        :returns: finalized order
        :rtype: messages.OrderResource

        """
        try:
            self.begin_finalization(orderr)
        except messages.Error as e:
            if e.code != 'orderNotReady':
                raise e
        return self.poll_finalization(orderr, deadline, fetch_alternative_chains)
```

Do you agree?

**Comment by frasertweedale** [comment]
> Do you agree?

Thanks for the PR.  I will review it next week.

**Comment by ohemorange** [comment]
Hey @frasertweedale, any thoughts on the proposed fix?

**Comment by wpodgorski** [comment]
@frasertweedale can you please review the fix provided by @ohemorange? This problem hit us hard and we are eagerly waiting to solve this and avoid any workarounds. Thank you in advance!
